File: mixtura/models/providers/homebrew/provider.py
"""
Homebrew package provider for Mixtura.

Provides integration with Homebrew package manager (macOS/Linux).
"""

import shutil
from typing import List, Optional

from mixtura.models.base import PackageManager
from mixtura.models.package import Package
from mixtura.cache import SearchCache


class HomebrewProvider(PackageManager):
    @property
    def name(self) -> str:
        return "homebrew"

    def is_available(self) -> bool:
        return shutil.which("brew") is not None

    def install(self, packages: List[str]) -> None:
        """
        Install packages via Homebrew.
        
        Raises:
            CommandError: If installation fails
        """
        if not self.is_available():
            raise RuntimeError("Homebrew is not installed.")

        self.run_command(["brew", "install"] + packages)

    def uninstall(self, packages: List[str]) -> None:
        """
        Remove Homebrew packages.
        
        Raises:
            CommandError: If uninstall fails
        """
        if not self.is_available():
            raise RuntimeError("Homebrew is not installed.")
        
        self.run_command(["brew", "uninstall"] + packages)

    def upgrade(self, packages: Optional[List[str]] = None) -> None:
        """
        Upgrade Homebrew packages.
        
        Raises:
            CommandError: If upgrade fails
        """
        if not self.is_available():
            raise RuntimeError("Homebrew is not installed.")

        if not packages:
            self.run_command(["brew", "upgrade"])
        else:
            self.run_command(["brew", "upgrade"] + packages)

    def list_packages(self) -> List[Package]:
        """Return list of installed Homebrew packages (installed on request only)."""
        if not self.is_available():
            return []

        try:
            # Get packages installed on request
            rc1, req_stdout, _ = self.run_capture(
                ["brew", "list", "--installed-on-request"]
            )
            if rc1 != 0:
                return []
            
            requested_pkgs = set(req_stdout.strip().split('\n'))
            requested_pkgs = {p.strip() for p in requested_pkgs if p.strip()}
            
            # Get versions
            rc2, ver_stdout, _ = self.run_capture(
                ["brew", "list", "--versions"]
            )
            
            if rc2 != 0:
                return []

            packages: List[Package] = []
            lines = ver_stdout.strip().split('\n')
            
            for line in lines:
                parts = line.strip().split()
                if len(parts) >= 2:
                    name = parts[0]
                    if name in requested_pkgs:
                        version = parts[1]
                        packages.append(Package(
                            name=name,
                            provider=self.name,
                            id=name,
                            version=version,
                            installed=True
                        ))

            return packages

        except Exception:
            return []

    def search(self, query: str) -> List[Package]:
        """Search for packages in Homebrew."""
        if not self.is_available():
            return []
        
        # Check cache first
        cache = SearchCache(self.name)
        cached = cache.get(query)
        if cached is not None:
            return cached
        
        try:
            returncode, stdout, stderr = self.run_capture(
                ["brew", "search", "--desc", query]
            )
             
            packages: List[Package] = []
            if returncode != 0 and not stdout:
                return []
             
            lines = stdout.strip().split('\n')
             
            current_type = "formula" 
             
            for line in lines:
                line = line.strip()
                if not line: continue
                if line.startswith("==> Formulae"):
                    current_type = "formula"
                    continue
                if line.startswith("==> Casks"):
                    current_type = "cask"
                    continue
                 
                if ": " in line:
                    parts = line.split(": ", 1)
                    name = parts[0]
                    desc = parts[1]
                else:
                    name = line
                    desc = "No description"
                 
                packages.append(Package(
                    name=name,
                    provider=self.name,
                    id=name,
                    version="unknown",
                    description=desc,
                    extra={"type": current_type}
                ))
             
            # Save to cache
            cache.set(query, packages)
            return packages

        except Exception:
            return []

    def clean(self) -> None:
        """
        Run Homebrew cleanup.
        
        Raises:
            CommandError: If cleanup fails
        """
        if not self.is_available():
            raise RuntimeError("Homebrew is not installed.")
        self.run_command(["brew", "cleanup"])

File: src/internal/provider/homebrew/provider.go
package homebrew

import (
	"strings"

	"mixtura/internal/core"
	"mixtura/internal/system"
)

type Provider struct{}

func (Provider) Name() string { return "homebrew" }

func (Provider) IsAvailable() bool { return system.CommandExists("brew") }

func (Provider) Install(packages []string) error {
	args := append([]string{"brew", "install"}, packages...)
	return system.RunCommand(args...)
}

func (Provider) Uninstall(packages []string) error {
	args := append([]string{"brew", "uninstall"}, packages...)
	return system.RunCommand(args...)
}

func (Provider) Upgrade(packages []string) error {
	args := []string{"brew", "upgrade"}
	args = append(args, packages...)
	return system.RunCommand(args...)
}

func (p Provider) ListPackages() []core.Package {
	if !p.IsAvailable() {
		return nil
	}

	rc, requestedOut, _ := system.CaptureCommand("brew", "list", "--installed-on-request")
	if rc != 0 {
		return nil
	}
	requested := map[string]bool{}
	for _, line := range strings.Split(requestedOut, "\n") {
		name := strings.TrimSpace(line)
		if name != "" {
			requested[name] = true
		}
	}

	rc, versionsOut, _ := system.CaptureCommand("brew", "list", "--versions")
	if rc != 0 {
		return nil
	}

	var packages []core.Package
	for _, line := range strings.Split(strings.TrimSpace(versionsOut), "\n") {
		parts := strings.Fields(line)
		if len(parts) < 2 || !requested[parts[0]] {
			continue
		}
		packages = append(packages, core.Package{
			Name:      parts[0],
			Provider:  p.Name(),
			ID:        parts[0],
			Version:   parts[1],
			Installed: true,
		})
	}
	return packages
}

func (p Provider) Search(query string) []core.Package {
	if !p.IsAvailable() {
		return nil
	}
	return system.CachedSearch(p.Name(), query, func() []core.Package {
		return p.searchUncached(query)
	})
}

func (p Provider) searchUncached(query string) []core.Package {
	rc, stdout, _ := system.CaptureCommand("brew", "search", "--desc", query)
	if rc != 0 && strings.TrimSpace(stdout) == "" {
		return nil
	}

	currentType := "formula"
	var packages []core.Package
	for _, line := range strings.Split(strings.TrimSpace(stdout), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		if strings.HasPrefix(line, "==> Formulae") {
			currentType = "formula"
			continue
		}
		if strings.HasPrefix(line, "==> Casks") {
			currentType = "cask"
			continue
		}

		name := line
		description := "No description"
		if parts := strings.SplitN(line, ": ", 2); len(parts) == 2 {
			name = parts[0]
			description = parts[1]
		}
		packages = append(packages, core.Package{
			Name:        name,
			Provider:    p.Name(),
			ID:          name,
			Version:     "unknown",
			Description: description,
			Extra:       map[string]string{"type": currentType},
		})
	}
	return packages
}

func (Provider) Clean() error {
	return system.RunCommand("brew", "cleanup")
}
